File: medium/0039_combination_sum/solution.cpp
#include <stack>
#include <tuple>
#include <vector>

class Solution {
public:
  std::vector<std::vector<int>> combinationSum(std::vector<int> &candidates,
                                               int target) {
    const size_t n = candidates.size();
    std::sort(candidates.begin(), candidates.end());

    std::stack<std::tuple<size_t, int, std::vector<int>>> s;
    s.push({0, 0, {}});

    std::vector<std::vector<int>> combinations;

    while (!s.empty()) {
      auto [i, sum, combination] = s.top();
      s.pop();

      if (sum == target) {
        combinations.push_back(combination);
        continue;
      }

      for (size_t j = i; j < n; ++j) {
        if (sum + candidates[j] <= target) {
          std::vector<int> c(combination);
          c.push_back(candidates[j]);
          s.push({j, sum + candidates[j], c});
        }
      }
    }

    return combinations;
  }
};

File: medium/0039_combination_sum/solution.py
class Solution:
    def combinationSum(
        self, candidates: list[int], target: int
    ) -> list[list[int]]:
        candidates.sort()

        combinations = []
        stack = [([], 0, 0)]

        while stack:
            curr_combination, curr_sum, index = stack.pop()

            if curr_sum == target:
                combinations.append(curr_combination)
                continue
            if curr_sum > target:
                continue

            for i in range(index, len(candidates)):
                if curr_sum + candidates[i] <= target:
                    stack.append(
                        (
                            curr_combination + [candidates[i]],
                            curr_sum + candidates[i],
                            i,
                        )
                    )

        return combinations
